almost_tcp/message_hdl.py

```diff
@@ -0,0 +1,249 @@
+"""
+Almost TCP data structures: messages and packet decoder.
+
+Almost TCP is a protocol designed to tunnel data streams
+(specifically, HTTP requests) over a serial port.
+It consists of a subset of TCP's functionality,
+including reduced sizes for some fields,
+with the goal of enabling a small hardware implementation.
+
+Each ATCP message consists of a fixed-length header
+followed by variable-length data.
+
+All numeric fields are network-endian (big-endian).
+
+-   Connection-oriented: syn/synack/ack setup
+-   Single stream number, not two port numbers
+    -   Stream number probing to determine # of hardware resources available
+
+"""
+
+from amaranth import Module, Signal, unsigned, Const
+from amaranth.lib.wiring import Component, In, Out, Signature
+from amaranth.lib import stream
+from amaranth.lib.data import UnionLayout, ArrayLayout, Struct
+from amaranth.lib.fifo import SyncFIFOBuffered
+
+
+class FlagsLayout(Struct):
+    """
+    Layout of the flags in an ATCP header.
+    """
+
+    # Indicates the sender has sent all the data it will send on this stream.
```

Comment:
Would slightly prefer if we used full words instead of 3-letter abbreviations, but this is OK if you like them.

Reply:
I didn't want to diverge from TCP, with the idea that eventually this might become a TCP implementation. [The spec](https://datatracker.ietf.org/doc/html/rfc9293#name-header-format) has descriptions for each of them, but the "name" is just the three letters.

(...I think so they fit nicely into the ASCII-art header format table, but, whatever.)